Use correct sizes for small images

mukeshpanchal27 opened the pull request:
## Summary

Fixes: #1251

For initial work:

- Only return a smaller image size if the content or wide size is larger than the image size.
   - If  content size is 620px and content use thumbnail the the size will use thumbnail 150px image size instead of 620px.
- Parse the width value for "px"; for other types( vw, min() ), return the image width according to the default WordPress behaviour.

<!--
For maintainers only, please make sure:

- PR has a `[Type]` label.
- PR has a plugin-specific milestone, or the `no milestone` label if it does not apply to any specific plugin.
- PR has a changelog-friendly title, or the `skip changelog` label if it should not be mentioned in the plugin's changelog.
-->

mukeshpanchal27 committed:
Use correct sizes for small images

joemcgill commented:
Thanks @mukeshpanchal27. It looks like the intent of this PR to constrain the `width` attribute value by the layout width. Is that correct? Is this related to an existing issue, or should a new issue be created to associate this PR with?

mukeshpanchal27 committed:
Fix merge conflict

mukeshpanchal27 committed:
Merge branch 'feature/more-accurate-sizes-attribute' into fix/small-image-sizes

mukeshpanchal27 committed:
Add image size tests coverage

mukeshpanchal27 committed:
Correct image urls

mukeshpanchal27 commented:
@joemcgill The PR is ready for review.

westonruter commented on plugins/auto-sizes/hooks.php:
```diff
@@ -91,6 +91,25 @@ function auto_sizes_render_generator(): void {
 add_action( 'wp_head', 'auto_sizes_render_generator' );
 
+/**
+ * Gets the smaller image size if the layout width id bigger.
```
```suggestion
 * Gets the smaller image size if the layout width is bigger.
```

westonruter commented on plugins/auto-sizes/hooks.php:
```diff
@@ -91,6 +91,25 @@ function auto_sizes_render_generator(): void {
+ * Gets the smaller image size if the layout width id bigger.
+ *
+ * It will return the smaller image size and return "px" if the layout width
+ * id something else. ex. min(640px, 90vw) or 90vw.
```
```suggestion
 * is something else, e.g. min(640px, 90vw) or 90vw.
```

westonruter commented on plugins/auto-sizes/hooks.php:
```diff
@@ -91,6 +91,25 @@ function auto_sizes_render_generator(): void {
+ * @return string The proper width after some calculations.
+ */
+function auto_sizes_get_width( string $layout_width, int $image_width ): string {
+	if ( 'px' === substr( $layout_width, -2 ) ) {
```
```suggestion
	if ( str_ends_with( $layout_width, 'px' ) {
```

joemcgill commented on plugins/auto-sizes/hooks.php:
```diff
@@ -107,9 +126,18 @@ function auto_sizes_improve_image_sizes_attribute( string $content, array $parse
+		$image_id   = $parsed_block['attrs']['id'] ?? '';
+		$image_size = $parsed_block['attrs']['sizeSlug'] ?? '';
+
+		$image_attributes = wp_get_attachment_image_src( $image_id, $image_size );
```
Instead of calling `wp_get_attachment_image_src()`, we should be able to parse the width attribute directly from the image markup using the existing `WP_HTML_Tag_Processor` instance. 

We should also factor in the `width` attribute from the parsed block, to cover cases where the image has been resized manually in the editor, which is another natural layout constraint.

mukeshpanchal27 replied:
@joemcgill, the editor does not add a width attribute in the image markup, so we can't retrieve it through `WP_HTML_Tag_Processor`.

Check the markup:
```
<!-- wp:image {"id":20,"sizeSlug":"large","linkDestination":"none"} -->
<figure class="wp-block-image size-large"><img src="http://localhost:8889/wp-content/uploads/2024/06/car-1024x576.jpeg" alt="" class="wp-image-20"/></figure>
<!-- /wp:image -->
```

In [this commit](https://github.com/WordPress/performance/pull/1252/commits/2656a0d355e8683ddb047cf8f5c5f06d1311aed0), I added support for image resizing (resizing allowed only in the default alignment).

mukeshpanchal27 committed:
Update docblock & use str_ends_with

mukeshpanchal27 committed:
Update code that account for resized image width

mukeshpanchal27 committed:
Update image size classes

joemcgill reviewed (changes requested):
After discussion with @mukeshpanchal27 and reviewing this PR in more detail, I've left the [following comment](https://github.com/WordPress/performance/issues/1251#issuecomment-2161443295) on the original issue with suggestions for a different approach that we should take to resolve this issue.

mukeshpanchal27 committed:
Exclude full and wide alignment from small size codebase

mukeshpanchal27 committed:
Fix spell

mukeshpanchal27 committed:
Revise the approach

mukeshpanchal27 committed:
Remove multiple parsing

mukeshpanchal27 commented:
@joemcgill @westonruter The PR is ready for review.

As discussed in [this comment](https://github.com/WordPress/performance/issues/1251#issuecomment-2161443295), I have updated the PR to add custom attributes to the `image` and `cover` blocks. These custom attributes are then removed in the `wp_content_img_tag` filter, which allows us to get both the alignment and image width.

@joemcgill  In [this commit](https://github.com/WordPress/performance/pull/1252/commits/4e56425dce141ae744424c32fb69b929a0f82e19), I have removed the multiple parsing and added an early check to skip if the custom attribute is not present.

joemcgill commented on plugins/auto-sizes/hooks.php:
```diff
@@ -100,16 +119,64 @@ function auto_sizes_render_generator(): void {
+		$has_auto_sizes = false;
+		if ( str_contains( $content, 'auto,' ) ) {
+			$has_auto_sizes = true;
+		}
```
Good idea, but we should be more specific here and only check the `sizes` attribute. I wonder if we avoid the need for this check by making sure this filter runs prior to the `auto_sizes_update_content_img_tag()` callback?

```suggestion
		$old_sizes      = $processor->get_attribute( 'sizes' );
		$has_auto_sizes = str_starts_with( $old_sizes, 'auto' );
```

mukeshpanchal27 replied:
Make sense. I have added higher priority for sizes filter and remove these code as now it not needed

joemcgill commented on plugins/auto-sizes/hooks.php:
```diff
@@ -100,16 +119,64 @@ function auto_sizes_render_generator(): void {
+ */
+function auto_sizes_improve_image_sizes_attributes( string $content ): string {
+	$processor = new WP_HTML_Tag_Processor( $content );
+	while ( $processor->next_tag( array( 'tag_name' => 'img' ) ) ) {
```
I think an `if` statement would be better here, rather than a while loop. I wouldn't ever expect the `wp_content_img_tag` filter to pass more than one `img` tag at a time, so we can bail early if `$processor->next_tag( array( 'tag_name' => 'img' )` returns `false`. Then you could return `$content` early if conditions aren't met.

Is there a particular use case you're thinking of here that I may be missing?

westonruter commented on plugins/auto-sizes/hooks.php:
```diff
@@ -100,16 +119,64 @@ function auto_sizes_render_generator(): void {
 
 	// Only update the markup if an image is found.
 	if ( $has_image ) {
+		$align = $parsed_block['attrs']['align'] ?? 'default';
+
+		if ( $align ) {
+			$processor->set_attribute( 'data-align', $align );
+		}
```
~Would `$align` ever be empty due to the `??`? If not, then this is simpler:~

```suggestion
		$align = parsed_block['attrs']['align'] ?? 'default';
		$processor->set_attribute( 'data-align', $align );
```

westonruter replied:
Alternatively, I see that this `data-align` attribute is serving two purposes: (1) exposing the `align` block attribute to the HTML tag processor and (2) indicating whether the tag has been processed. What if instead the tag processor added a unique attribute like `data-needs-sizes-update` and that this attribute is then looked for when avoiding redundant optimization. So something like this:

```suggestion
		$processor->set_attribute( 'data-needs-sizes-update', true );

		$align = $parsed_block['attrs']['align'] ?? '';
		if ( $align ) {
			$processor->set_attribute( 'data-align', $align );
		}
```

And then see suggestions below.

joemcgill commented on plugins/auto-sizes/hooks.php:
```diff
@@ -100,16 +119,64 @@ function auto_sizes_render_generator(): void {
 		$layout = wp_get_global_settings( array( 'layout' ) );
-		$align  = $parsed_block['attrs']['align'] ?? null;
-		$sizes  = null;
+		$align  = $processor->get_attribute( 'data-align' );
```
This duplicates the `$image_alignment` variable above and can be removed.

westonruter commented on plugins/auto-sizes/hooks.php:
```diff
@@ -100,16 +119,64 @@ function auto_sizes_render_generator(): void {
+			continue; // Skip if width attribute is not found.
+		}
+
+		// Skips second time parsing if custom IMG attribute removed.
+		$image_alignment = $processor->get_attribute( 'data-align' );
+		if ( ! $image_alignment ) {
+			continue;
+		}
```
```suggestion
		// Skips second time parsing if already processed.
		if ( false === $processor->get_attribute( 'data-needs-sizes-update' ) ) {
			continue;
		}
```

westonruter commented on plugins/auto-sizes/hooks.php:
```diff
@@ -125,18 +192,23 @@ function auto_sizes_improve_image_sizes_attribute( string $content, array $parse
 		}
 
-		$content = $processor->get_updated_html();
+		$processor->remove_attribute( 'data-align' );
```
```suggestion
		$processor->remove_attribute( 'data-needs-sizes-update' );
		$processor->remove_attribute( 'data-align' );
```

mukeshpanchal27 replied:
PR updated with the changes.

joemcgill commented on plugins/auto-sizes/hooks.php:
```diff
@@ -100,16 +119,64 @@ function auto_sizes_render_generator(): void {
+		$image_width = $processor->get_attribute( 'width' );
+		if ( ! $image_width ) {
+			continue; // Skip if width attribute is not found.
+		}
```
`$image_width` is only a requirement for default aligned images. We could still update full/wide aligned images even if no `width` attribute is present.

joemcgill commented on plugins/auto-sizes/tests/improve-sizes-test.php:
```diff
@@ -24,85 +24,198 @@ public static function set_up_before_class(): void {
 
 		self::$image_id = self::factory()->attachment->create_upload_object( TESTS_PLUGIN_DIR . '/tests/data/images/leaves.jpg' );
+
+		// Disable lazy loading attribute.
+		add_filter( 'wp_img_tag_add_loading_attr', '__return_false' );
```
This is a risky filter to disable while running these tests, because we'll be missing scenarios where the loading attribute changes the way the markup gets created. I assume you did this to avoid having auto sizes added while running these tests? If so, instead of removing the `wp_img_tag_add_loading_attr` filter, I would suggest running these without the auto-sizes filter, which seems to be what is causing potential failures here.
```suggestion
		// Disable auto sizes in these tests.
		remove_filter( 'wp_content_img_tag', 'auto_sizes_update_content_img_tag' );
```

However, since the hooks global won't get reset after this class runs, you'll also need to add this hook back after the class runs, or else it will affect any tests that run afterwards. Ex:

```php
	public static function tear_down_after_class(): void {
		parent::tear_down_after_class();

		add_filter( 'wp_content_img_tag', 'auto_sizes_update_content_img_tag' );
	}
```

If there are specific tests where you need to disable lazy-loading, it would be better to modify hooks directly in that test, because then they will automatically get reset during the `WP_UnitTestCase_Base::tear_down()` method that runs after each test.

mukeshpanchal27 replied:
Thanks @joemcgill for suggestion. This will not working as the filter is re-added through `the_content` filter so it show error in tests.

```
1) Tests_Improve_Sizes::test_image_block_with_full_alignment with data set "Return wideSize 1280px instead of medium size 300px" ('medium')
Failed asserting that '<figure class="wp-block-image size-medium"><img loading="lazy" decoding="async" width="300" height="200"   src="http://localhost:8889/wp-content/uploads/2024/06/leaves-283-300x200.jpg" alt="" class="wp-image-5" srcset="http://localhost:8889/wp-content/uploads/2024/06/leaves-283-300x200.jpg 300w, http://localhost:8889/wp-content/uploads/2024/06/leaves-283-1024x683.jpg 1024w, http://localhost:8889/wp-content/uploads/2024/06/leaves-283-768x512.jpg 768w, http://localhost:8889/wp-content/uploads/2024/06/leaves-283.jpg 1080w" sizes="auto, 100vw" /></figure>' contains "sizes="100vw" ".

/var/www/html/wp-content/plugins/performance/plugins/auto-sizes/tests/test-improve-sizes.php:53

2) Tests_Improve_Sizes::test_image_block_with_full_alignment with data set "Return wideSize 1280px instead of large size 1024px" ('large')
Failed asserting that '<figure class="wp-block-image size-large"><img loading="lazy" decoding="async" width="1024" height="683"   src="http://localhost:8889/wp-content/uploads/2024/06/leaves-283-1024x683.jpg" alt="" class="wp-image-5" srcset="http://localhost:8889/wp-content/uploads/2024/06/leaves-283-1024x683.jpg 1024w, http://localhost:8889/wp-content/uploads/2024/06/leaves-283-300x200.jpg 300w, http://localhost:8889/wp-content/uploads/2024/06/leaves-283-768x512.jpg 768w, http://localhost:8889/wp-content/uploads/2024/06/leaves-283.jpg 1080w" sizes="auto, 100vw" /></figure>' contains "sizes="100vw" ".

/var/www/html/wp-content/plugins/performance/plugins/auto-sizes/tests/test-improve-sizes.php:53

3) Tests_Improve_Sizes::test_image_block_with_full_alignment with data set "Return wideSize 1280px instead of full size 1080px" ('full')
Failed asserting that '<figure class="wp-block-image size-full"><img loading="lazy" decoding="async" width="1080" height="720"   src="http://localhost:8889/wp-content/uploads/2024/06/leaves-283.jpg" alt="" class="wp-image-5" srcset="http://localhost:8889/wp-content/uploads/2024/06/leaves-283.jpg 1080w, http://localhost:8889/wp-content/uploads/2024/06/leaves-283-300x200.jpg 300w, http://localhost:8889/wp-content/uploads/2024/06/leaves-283-1024x683.jpg 1024w, http://localhost:8889/wp-content/uploads/2024/06/leaves-283-768x512.jpg 768w" sizes="auto, 100vw" /></figure>' contains "sizes="100vw" ".

/var/www/html/wp-content/plugins/performance/plugins/auto-sizes/tests/test-improve-sizes.php:53

```

joemcgill commented on plugins/auto-sizes/tests/improve-sizes-test.php:
```diff
@@ -24,85 +24,198 @@ public static function set_up_before_class(): void {
-		$parsed_block = array(
-			'attrs' => array(
-				// No alignment specified.
+	public function data_image_sizes_for_wide_alignment(): array {
+		return array(
+			'Return wideSize 1280px instead of thumb size 150px'  => array(
+				'thumbnail',
+			),
+			'Return wideSize 1280px instead of medium size 300px'  => array(
+				'medium',
+			),
+			'Return wideSize 1280px instead of large size 1024px'  => array(
+				'large',
+			),
+			'Return wideSize 1280px instead of full size 1080px'  => array(
+				'full',
 			),
 		);
+	}
```
This data provider is essentially returning the same values as `data_image_sizes()`. Do we need both?

joemcgill commented on plugins/auto-sizes/tests/improve-sizes-test.php:
```diff
@@ -24,85 +24,198 @@ public static function set_up_before_class(): void {
+	 * @param string $expected   Expected output.
+	 * @param string $is_resize  Whether resize or not.
+	 */
+	public function test_image_block_with_default_alignment( string $image_size, string $expected, string $is_resize = '' ): void {
```
I'd pass a `bool` to `$is_resize`, rather than a string. Will need to update the data provider cases to pass `true` instead of 'yes'.

joemcgill reviewed (changes requested):
I've tested locally, and this seems to be working as expected for most of the use cases we're trying to solve with this PR. I've left some code quality feedback that would be good to address before merging.

mukeshpanchal27 committed:
Fix phpstan issue

mukeshpanchal27 committed:
Fix unit tests

mukeshpanchal27 committed:
Load filter prior to auto sizes filter

mukeshpanchal27 committed:
Address review feedbacks

mukeshpanchal27 committed:
Address unit tests feedback

mukeshpanchal27 committed:
Update file name for consistency

mukeshpanchal27 committed:
Fix unit tests

mukeshpanchal27 committed:
Clean-up

mukeshpanchal27 commented:
@joemcgill @westonruter Thanks for the feedback. I have updated the PR as per your directions. It is now ready for review.

westonruter committed:
Merge branch 'trunk' of https://github.com/WordPress/performance into fix/small-image-sizes

* 'trunk' of https://github.com/WordPress/performance: (113 commits)
  Further refine command explaining type casting
  Elaborate on comment explaining why type casting is added
  Add try/catch block around imagecolorsforindex()
  Fix grammar typos
  Fix static analysis issues in Test_WebP_Uploads_Image_Edit
  Eliminate checks for seemingly impossible false return values
  Update PHPStan to 1.11.5
  Add missing '@'
  Apply suggestions from code review
  Update testcase
  Fix wp:gallery styling not applying
  Update banner image
  Extend check code for WordPress 6.6
  Harmonize return descriptions
  Bump Optimization Detective 0.3.1
  Log URL metrics group collection to console when WP_DEBUG
  Include non-intersecting elements in URL metrics
  Bump version to 0.1.1
  Fix logic error for preg_match() return value
  Add failing test case for a non-background image style
  ...

westonruter committed:
Merge branch 'feature/more-accurate-sizes-attribute' into fix/small-image-sizes

* feature/more-accurate-sizes-attribute:

westonruter reviewed (approved):
LGTM

joemcgill committed:
Revert changes to composer.lock file

joemcgill committed:
Properly revert composer.lock file

joemcgill commented on plugins/auto-sizes/hooks.php:
```diff
@@ -100,43 +119,93 @@ function auto_sizes_render_generator(): void {
+
+	return $processor->get_updated_html();
+}
+// Run filter prior to auto sizes "auto_sizes_update_content_img_tag" filter.
```
💖

joemcgill reviewed (approved):
This is working as expected. I reverted the changes to the `composer.lock` file that were added because I think they were unintentional and doesn't seem to be related to this change. If that was an incorrect assumption we can update composer dependencies in a follow-up task.
